aquadoggo/src/db/stores/task.rs

```diff
@@ -0,0 +1,210 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use anyhow::Result;
+use p2panda_rs::document::{DocumentId, DocumentViewId};
+use sqlx::{query, query_as};
+
+use crate::db::errors::SqlStorageError;
+use crate::db::models::TaskRow;
+use crate::db::provider::SqlStorage;
+use crate::materializer::{Task, TaskInput};
+
+/// Methods to interact with the `tasks` table in the database.
+impl SqlStorage {
+    /// Inserts a "pending" task into the database.
+    pub async fn insert_task(&self, task: &Task<TaskInput>) -> Result<(), SqlStorageError> {
+        // Convert task input to correct database types
+        let task_input = task.input();
+        let document_id = task_input.document_id.as_ref().map(|id| id.as_str());
+        let document_view_id = task_input
+            .document_view_id
+            .as_ref()
+            .map(|view_id| view_id.as_str());
+
+        // Check first if this task already exists, to avoid duplicate rows
+        let task_row = query_as::<_, TaskRow>(
+            "
+            SELECT
+                name,
+                document_id,
+                document_view_id
+            FROM
+                tasks
+            ",
+        )
+        .fetch_optional(&self.pool)
+        .await
+        .map_err(|err| SqlStorageError::Transaction(err.to_string()))?;
+
+        // If yes, we are already done here
+        if task_row.is_some() {
+            return Ok(());
+        }
+
+        // Insert task into database
+        let result = query(
+            "
+            INSERT INTO
+                tasks (
+                    name,
+                    document_id,
+                    document_view_id
+                )
+            VALUES
+                ($1, $2, $3)
+            ",
+        )
+        .bind(task.worker_name())
+        .bind(document_id)
+        .bind(document_view_id)
+        .execute(&self.pool)
+        .await
+        .map_err(|err| SqlStorageError::Transaction(err.to_string()))?;
+
+        if result.rows_affected() != 1 {
+            Err(SqlStorageError::Insertion("tasks".into()))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Removes a "pending" task from the database.
+    pub async fn remove_task(&self, task: &Task<TaskInput>) -> Result<(), SqlStorageError> {
+        // Convert task input to correct database types
+        let task_input = task.input();
+        let document_id = task_input.document_id.as_ref().map(|id| id.as_str());
+        let document_view_id = task_input
+            .document_view_id
+            .as_ref()
+            .map(|view_id| view_id.as_str());
+
+        // Remove task from database
+        let result = query(
+            "
+            DELETE FROM
+                tasks
+            WHERE
+                name IS $1
```

Comment:
I've never seen IS used for comparison and couldn't find anything in the pg manual. Does it actually work? The standard operator is = and I only know IS as part of sth like IS DISTINCT.

Reply:
Yes, I know! Funnily it didn't work when I did `=` .. and using `IS` it suddenly worked. Maybe because there are `NULL` values involved?

Reply:
Indeed, this works:

```
            DELETE FROM
                tasks
            WHERE
                name = $1
                AND document_id IS $2
                AND document_view_id IS $3
```

And changing all to `=` fails as `document_id` and `document_view_id` can be `NULL` and we need to exactly match that to fullfil the query.